Add unit tests from lldb-eval

kuilpd opened the pull request:
Add unit tests ported from lldb-eval.

TODO:
* Fix errors in XFailed tests
* Fix segfault errors in GTEST_SKIP tests
* Implement interfaces for scoped, context, and separate parsing and re-enable DISABLED_ tests

kuilpd committed:
Add a fuzzer

Build by adding '-DLLVM_USE_SANITIZER=Address -DLLVM_USE_SANITIZE_COVERAGE=On' to the CMake configuration.

kuilpd committed:
Get fuzzer test binary via a relative path

kuilpd committed:
Add unittests from lldb-eval

Removed for now:
* All tests with context variables evaluation: need to implement an call interface to DIL
* All tests with evaluation with scope: need to implement an call interface to DIL
* Some tests that crash the test suite completely: some of the tests that involve nullptr

kuilpd committed:
Add an API call to DIL from SBFrame

kuilpd commented:
Since not all test cases were moved to python tests, I added unit tests from lldb-eval to see better what functionality is missing. Notably, a lot of expressions are not making it past the lexer now, many of which can be fixed by solving issue #4 

This pull request is just to show the branch. My question is, should we maintain these unit tests or move them one by one to python suite as we fix them?

kuilpd committed:
Change 'aka' to 'canonically referred to' in error messages

kuilpd committed:
Add more cases for floating numbers

kuilpd committed:
Remove fuzzer (will move to another branch)

kuilpd committed:
Apply git-clang-format

kuilpd committed:
Adjust division by zero tests to check the error message

kuilpd committed:
Add XFail matcher with a counter & mark failed tests XFail

kuilpd committed:
Disable tests suites with registers, chars, strings

kuilpd committed:
Skip crashing nullptr tests & restore working nullptr tests

kuilpd committed:
Create placeholders for scope, context, parsing evaluation & disable relative tests

kuilpd committed:
Restore TestStaticConstDeclaredInline

kuilpd committed:
Fix compiler warnings

kuilpd committed:
DIL TODO comment

kuilpd committed:
Remove libstdc++ test binary

kuilpd commented:
@cmtice
Cleaned up the unit tests:
* Marked currently failing tests as XFail and they will pass until the issue is fixed (every test suite will print the number of these tests in the summary)
* Skipped 3 tests that cause a segfault, marked by GTEST_SKIP
* Added placeholders for currently unimplemented functionality: scope evaluation, evaluation with context, separate parsing. They return an error when running a relevant test
* Disabled all the test that use unimplemented functionality, marked by DISABLED_ and printed out at the end of unit test summary

Haven't figured out yet how to build the test binary from source and pass it to the unit tests, so for now it's still a pre-built binary.
If it doesn't work on your PC, `clang++ -O0 -g -stdlib=libc++ test_binary.cc -o test_binary_libc++.bin` in `lldb/unittests/DIL/Inputs` directory will rebuild it for your system.

Please try launching the tests and tell me how it goes :)
If it works decent enough, we can merge it to main and then fix the tests along with bugs in other pull requests.

kuilpd committed:
Compile the test binary with the in-tree clang and lld

kuilpd commented:
@cmtice 
Added a way to compile the test binary with the in-tree `clang`. It's kind of hacky, requires `clang` and `lld` projects enabled, and `libc++` installed in the system (some of the tests depend specifically on being linked against libc++, I will probably adjust that later)

cmtice commented on lldb/unittests/DIL/Inputs/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+get_target_property(EXE_PATH LLDBDILTests RUNTIME_OUTPUT_DIRECTORY)
+add_custom_command(
+    OUTPUT test_binary.bin
+    COMMAND $<TARGET_FILE:clang> test_binary.cc -std=c++17 -fuse-ld=lld -B$<TARGET_FILE:lld> 
```
```suggestion
    COMMAND $<TARGET_FILE:clang> ${CMAKE_CURRENT_SOURCE_DIR}/test_binary.cc -std=c++17 -fuse-ld=lld -B$<TARGET_FILE:lld> 
```

cmtice replied:
I found that I needed to tell clang what directory to look in, for building test_binary.cc

cmtice commented:
I found that with the small change I've shown above, the unittest builds & runs for me.  I get 2 failures: 
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] EvalTest.TestCxxReinterpretCast
[  FAILED  ] EvalTest.TestUniquePtr

kuilpd committed:
Add the full path to test_binary.cc

kuilpd commented:
Hmm, these are the tests that fail when I link against libstdc++.
I should add checks for which library is used anyway, will do that next.

kuilpd committed:
Link test_binary against in-tree libc++

kuilpd committed:
Make unique_ptr tests work with current libc++

kuilpd commented:
@cmtice 
Added linking against in-tree libc++ and adjusted tests accordingly, so that there is no ambiguity about which library is used.
Just need to enable runtimes `libcxx;libcxxabi;libunwind` in the project.

kuilpd committed:
Rename files

kuilpd commented:
@cmtice Could you please check if the tests work after the last changes?

cmtice commented:
> @cmtice Could you please check if the tests work after the last changes?

The tests build just fine. When I run them I get the following results:
[  PASSED  ] 60 tests.
[  SKIPPED ] 3 tests, listed below:
[  SKIPPED ] EvalTest.TestCStyleCastPointer
[  SKIPPED ] EvalTest.TestCxxStaticCast
[  SKIPPED ] EvalTest.TestTernaryOperator
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EvalTest.TestCxxReinterpretCast

 1 FAILED TEST
  YOU HAVE 23 DISABLED TESTS

kuilpd commented:
Okay, disabled and skipped tests are supposed to be there, but nothing should fail. What are the errors?

cmtice commented:
> Okay, disabled and skipped tests are supposed to be there, but nothing should fail. What are the errors?

[ RUN      ] EvalTest.TestCxxReinterpretCast
/usr/local/google3/cmtice/access-softek-dil/lldb/unittests/DIL/DILTests.cpp:1779: Failure
Value of: Eval("reinterpret_cast<nullptr_t>(ptr)")
Expected: evaluates with an error: 'reinterpret_cast from 'int *' to 'nullptr_t' (canonically referred to as 'std::nullptr_t') is not allowed'
  Actual: { DIL: NULL, lldb: NULL }, error:
<expr>:1:18: unknown type name 'nullptr_t'
reinterpret_cast<nullptr_t>(ptr)
                 ^              

/usr/local/google3/cmtice/access-softek-dil/lldb/unittests/DIL/DILTests.cpp:1783: Failure
Value of: Eval("reinterpret_cast<nullptr_t>(0)")
Expected: evaluates with an error: 'reinterpret_cast from 'int' to 'nullptr_t' (canonically referred to as 'std::nullptr_t') is not allowed'
  Actual: { DIL: NULL, lldb: NULL }, error:
<expr>:1:18: unknown type name 'nullptr_t'
reinterpret_cast<nullptr_t>(0)
                 ^            

/usr/local/google3/cmtice/access-softek-dil/lldb/unittests/DIL/DILTests.cpp:1787: Failure
Value of: Eval("reinterpret_cast<nullptr_t>(nullptr)")
Expected: evaluates with an error: 'reinterpret_cast from 'std::nullptr_t' to 'nullptr_t' (canonically referred to as 'std::nullptr_t') is not allowed'
  Actual: { DIL: NULL, lldb: NULL }, error:
<expr>:1:18: unknown type name 'nullptr_t'
reinterpret_cast<nullptr_t>(nullptr)
                 ^

kuilpd committed:
Skip 3 tests with error string mismatch

kuilpd commented:
I skipped these tests for now, it's just the error message not matching exactly, will figure it out later why do they match on my build, but not on yours.
If none of the tests fail now, I guess we can merge?

cmtice commented:
> I skipped these tests for now, it's just the error message not matching exactly, will figure it out later why do they match on my build, but not on yours. If none of the tests fail now, I guess we can merge?

SGTM!

kuilpd committed:
Move GTEST_SKIP() to the end of test fixtures

kuilpd commented:
I misunderstood how `GTEST_SKIP()` works, it skips the rest of the test fixture and marks it skipped, so I moved the tests I want to skip to the end of their fixture.
